### PR TITLE
feat(blockingproxy): Add synchronization with BlockingProxy.

### PR DESCRIPTION
**Pull request by mgiambalvo**
This adds support for BlockingProxy behind the flag --useBlockingProxy. 

If set, the driver providers will start a proxy during their setup phase, passing the selenium address to the proxy and starting a webdriver client that talks to the proxy.

Starting a proxy for each driver provider isn't strictly necessary. However, when we run with multiple capabilities it's easier to handle the logging if each Protractor instance has it's own proxy. 

Known issues:
- Doesn't work with directConnect. You can get the address of chromedriver by mucking around in Selenium internals, but this probably changed for Selenium 3.0 and I doubt it's worth figuring out until we upgrade.
- Doesn't yet work with webDriverProxy (but it's an easy fix)
- Doesn't work with wrapDriver. The new browser instance will need to know somehow that the flag is set, still working on this.

### CODE REVIEW

File: lib/browser.ts [hunk 1]
```diff
@@ -188,7 +195,21 @@ export class ProtractorBrowser extends Webdriver {
    *
    * @type {boolean}
    */
-  ignoreSynchronization: boolean;
+  set ignoreSynchronization(value) {
```
Comment (juliemr) [inline lib/browser.ts]: I think we should take this opportunity to improve the naming here.  I think we had discussed `waitForAngularEnabled(true/false)`

Comment (juliemr) [inline lib/browser.ts]: I'd also be OK with keeping these changes that keep `ignoreSynchronization` working but adding the new API as the 'recommended way' as well.

File: lib/browser.ts [hunk 2]
```diff
@@ -668,6 +693,12 @@ export class ProtractorBrowser extends Webdriver {
       return 'Protractor.get(' + destination + ') - ' + str;
     };
 
+    if (this.bpClient) {
```
Comment (juliemr) [inline lib/browser.ts]: Hm, why is this necessary? Because we do executeScript calls, and BpClient isn't smart enough  to know that it doesn't need to wait for those?

File: lib/config.ts [hunk 3]
```diff
@@ -88,6 +88,13 @@ export interface Config {
    */
   webDriverProxy?: string;
 
+  /**
+   * If specified, connect to webdriver through a proxy that manages client-side
+   * synchronization. This option is intended for testing an  experimental feature
+   * and is not meant to be used by end users.
```
Comment (juliemr) [inline lib/config.ts]: This is a bit scary - maybe just 

"Blocking Proxy is an experimental feature and may change without notice."?

Also, I think this should be boolean?

File: lib/driverProviders/driverProvider.ts [hunk 4]
```diff
@@ -28,17 +31,28 @@ export class DriverProvider {
     return this.drivers_.slice();  // Create a shallow copy
   }
 
+  getBPUrl() {
+    return `http://localhost:${this.bpRunner.port}`;
```
Comment (juliemr) [inline lib/driverProviders/driverProvider.ts]: use os.hostname()?

File: lib/driverProviders/driverProvider.ts [hunk 5]
```diff
@@ -89,22 +103,33 @@ export class DriverProvider {
...
   /**
-   * Default setup environment method.
-   * @return a promise
+   * Default setup environment method, common to all driver providers.
    */
   setupEnv(): q.Promise<any> {
-    return q.fcall(function() {});
+    let driverPromise = this.setupDriverEnv();
+    if (this.config_.useBlockingProxy) {
+      // TODO(heathkit): If set, pass the webDriverProxy to BP.
```
Comment (juliemr) [inline lib/driverProviders/driverProvider.ts]: You should just grab this github username :)

File: lib/driverProviders/driverProvider.ts [hunk 6]
```diff
@@ -89,22 +103,33 @@ export class DriverProvider {
...
+      return q.all([driverPromise, this.bpRunner.start()]);
+    }
+    return driverPromise;
   };
 
   /**
+   * Set up environment specific to a particular driver provider. Overridden
+   * by each driver provider.
+   */
+  protected abstract setupDriverEnv():
```
Comment (juliemr) [inline lib/driverProviders/driverProvider.ts]: Also update the documentation in driverProviders/README.md

File: lib/driverProviders/driverProvider.ts [hunk 7]
```diff
@@ -89,22 +103,33 @@ export class DriverProvider {
...
+      /**
    * Teardown and destroy the environment and do any associated cleanup.
    * Shuts down the drivers.
    *
    * @public
    * @return {q.promise} A promise which will resolve when the environment
    *     is down.
    */
-  teardownEnv(): q.Promise<q.Promise<webdriver.WebDriver>[]> {
+      teardownEnv(): q.Promise<q.Promise<webdriver.WebDriver>[]> {
```
Comment (juliemr) [inline lib/driverProviders/driverProvider.ts]: This extra spacing doesn't look right...

File: package.json [hunk 8]
```diff
@@ -72,7 +73,7 @@
   },
   "license": "MIT",
   "engines": {
-    "node": ">=4.2.x"
+    "node": ">=6.9.x"
```
Comment (juliemr) [inline package.json]: Let's avoid this change until 5.x

Out of curiosity, what in this PR depends on it?

File: spec/blockingProxyConf.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
+var env = require('./environment.js');
```
Comment (juliemr) [inline spec/blockingProxyConf.js]: Add this to the test.sh script for `npm test`?

File: spec/blockingProxyConf.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+// The main suite of Protractor tests.
```
Comment (juliemr) [inline spec/blockingProxyConf.js]: Update comment.

File: spec/blockingProxyConf.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  specs: [
+    'basic/synchronize_spec.js',
+    'ng2/async_spec.js',
+    'ng2/timeout_spec.js',
+    'hybrid/async_spec.js'
+  ],
+
+  blockingProxyUrl: 'http://localhost:8111',
+
+  // Exclude patterns are relative to this directory.
```
Comment (juliemr) [inline spec/blockingProxyConf.js]: Unnecessary.

File: spec/blockingProxyConf.js [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+  // Spec patterns are relative to this directory.
+  specs: [
+    'basic/synchronize_spec.js',
+    'ng2/async_spec.js',
+    'ng2/timeout_spec.js',
+    'hybrid/async_spec.js'
+  ],
+
+  blockingProxyUrl: 'http://localhost:8111',
```
Comment (juliemr) [inline spec/blockingProxyConf.js]: I think this should be `useBlockingProxy: true` now

File: lib/browser.ts (hunk 1 above)
Comment (mgiambalvo) [inline lib/browser.ts]: Good point. I'll add the new API, but I'd like to leave `ignoreSynchronization` in place and announce that it's deprecated.

File: lib/browser.ts (hunk 2 above)
Comment (mgiambalvo) [inline lib/browser.ts]: Exactly. We could skip waiting for executeScript calls in general, but I wanted to keep behavior the same.

File: lib/config.ts (hunk 3 above)
Comment (mgiambalvo) [inline lib/config.ts]: Done

File: lib/driverProviders/driverProvider.ts (hunk 4 above)
Comment (mgiambalvo) [inline lib/driverProviders/driverProvider.ts]: Done

File: lib/bpRunner.ts [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+      })
+    })
+  }
+
+  checkSupportedConfig() {
+    if (this.config.directConnect) {
+      throw new Error('BlockingProxy not yet supported with directConnect!');
+    }
+  }
+}
```
Comment (mgiambalvo) [inline lib/bpRunner.ts]: Done

File: lib/bpRunner.ts [hunk 14]
```diff
@@ -0,0 +1,62 @@
+import {ChildProcess, fork} from 'child_process';
+import * as q from 'q';
+
+import {Config} from './config';
+import {Logger} from './logger';
+
+const BP_PATH = 'node_modules/blocking-proxy/built/lib/bin.js';
```
Comment (mgiambalvo) [inline lib/bpRunner.ts]: Done

File: package.json (hunk 8 above)
Comment (mgiambalvo) [inline package.json]: Rebased to the beta branch.

File: lib/driverProviders/driverProvider.ts (hunk 7 above)
Comment (mgiambalvo) [inline lib/driverProviders/driverProvider.ts]: That is weird, don't know why clang-format didn't catch this...

File: lib/driverProviders/driverProvider.ts (hunk 4 above)
Comment (mgiambalvo) [inline lib/driverProviders/driverProvider.ts]: I think localhost is the right thing to do here, since BP is binding to localhost.

File: lib/driverProviders/driverProvider.ts (hunk 5 above)
Comment (mgiambalvo) [inline lib/driverProviders/driverProvider.ts]: Yup

**Comment by juliemr** [comment]
Failing test is the one we decided to kill - so after this is gone and passing, LGTM.

**Comment by juliemr** [comment]
Dunno why CLA-bot is so unhappy today.
